sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosClientCache.scala

```diff
@@ -129,152 +130,196 @@ private[spark] object CosmosClientCache extends BasicLoggingTrait {
+        val cosmosAsyncClient = createCosmosAsyncClient(cosmosClientConfiguration, cosmosClientStateHandle)
+        var sparkCatalogClient: CosmosCatalogClient = CosmosCatalogCosmosSDKClient(cosmosAsyncClient)
+        // When using AAD auth, cosmos catalog will change to use management sdk instead of cosmos sdk
+        cosmosClientConfiguration.authConfig match {
+            case aadAuthConfig: CosmosAadAuthConfig =>
+                sparkCatalogClient =
+                    CosmosCatalogManagementSDKClient(
+                        aadAuthConfig.resourceGroupName,
+                        aadAuthConfig.databaseAccountName,
+                        createCosmosManagementClient(aadAuthConfig),
+                        cosmosAsyncClient)
+            case _ =>
+        }
+
+        val epochNowInMs = Instant.now.toEpochMilli
+        val owners = new TrieMap[OwnerInfo, Option[Boolean]]
+        owners.put(ownerInfo, None)
+
+        val newClientCacheEntry = CosmosClientCacheMetadata(
+          cosmosAsyncClient,
+          sparkCatalogClient,
+          cosmosClientConfiguration,
+          new AtomicLong(epochNowInMs),
+          new AtomicLong(epochNowInMs),
+          new AtomicLong(epochNowInMs),
+          new AtomicLong(1),
+          owners)
+
+        cache.putIfAbsent(clientConfigWrapper, newClientCacheEntry) match {
+          case None => new CacheItemImpl(cosmosAsyncClient, sparkCatalogClient, newClientCacheEntry, ownerInfo)
+          case Some(_) =>
+            throw new ConcurrentModificationException("Should not reach here because its synchronized")
+        }
+    }
+  }
+
+  // scalastyle:off method.length
+  // scalastyle:off cyclomatic.complexity
+  private[this] def createCosmosAsyncClient(cosmosClientConfiguration: CosmosClientConfiguration,
+                                            cosmosClientStateHandle: Option[CosmosClientMetadataCachesSnapshot]): CosmosAsyncClient = {
+      var builder = new CosmosClientBuilder()
           .endpoint(cosmosClientConfiguration.endpoint)
           .userAgentSuffix(cosmosClientConfiguration.applicationName)
           .throttlingRetryOptions(
-            new ThrottlingRetryOptions()
-              .setMaxRetryAttemptsOnThrottledRequests(Int.MaxValue)
-              .setMaxRetryWaitTime(Duration.ofSeconds((Integer.MAX_VALUE/1000) - 1)))
+              new ThrottlingRetryOptions()
+                  .setMaxRetryAttemptsOnThrottledRequests(Int.MaxValue)
+                  .setMaxRetryWaitTime(Duration.ofSeconds((Integer.MAX_VALUE / 1000) - 1)))
+
+      val authConfig = cosmosClientConfiguration.authConfig
+      authConfig match {
+          case masterKeyAuthConfig: CosmosMasterKeyAuthConfig => builder.key(masterKeyAuthConfig.accountKey)
+          case aadAuthConfig: CosmosAadAuthConfig =>
+              val tokenCredential = new ClientSecretCredentialBuilder()
+                  .authorityHost(aadAuthConfig.authorityHost)
+                  .tenantId(aadAuthConfig.tenantId)
```

Comment:
These configs I get - why did we also include SubscriptionId and ResourceGroup in the settings?

Reply:
They are needed here
```
  private[this] def createCosmosManagementClient(authConfig: CosmosAadAuthConfig): CosmosManager = {
      val azureProfile = new AzureProfile(authConfig.tenantId, authConfig.subscriptionId, authConfig.azureEnvironment)
      val tokenCredential = new ClientSecretCredentialBuilder()
          .authorityHost(authConfig.azureEnvironment.getActiveDirectoryEndpoint())
          .tenantId(authConfig.tenantId)
          .clientId(authConfig.clientId)
          .clientSecret(authConfig.clientSecret)
          .build()
      CosmosManager.authenticate(tokenCredential, azureProfile)
  }
```

Reply:
Discussed offline, since resourceGroupName is a required parameters in ManagementSDK, so will expose it. Will follow up offline with Management SDK team